Fix the result structure of DescribeInputRewrite and DescribeOutputRewrite

goldmedal opened the pull request:
I found the result of `DescribeInputRewrite` and `DescribeOutputRewrite` is wrong. If using `SqlFormatter` to format their result, the syntax of results is wrong.

For example:
1. Prepare a statement `select comment from tpch.tiny.orders` and name it `q1`.
2. Rewrite the sql `DESCRIBE OUTPUT q1`.
3. Formatted the rewrite result to SQL using `SqlFormatter`, and then we will get:
```sql
SELECT
  "Column Name"
, Catalog
, Schema
, Table
, Type
, "Type Size"
, Aliased
FROM
   VALUES 
  ROW ('comment', 'tpch', 'tiny', 'orders', 'varchar(79)', 0, false)
 "Statement Output" ("Column Name", Catalog, Schema, Table, Type, "Type Size", Aliased)
```

This SQL exists two syntax errors.
1. The name of some identifier are the reserved word of trino, such as `Catalog`, `Schema` and `Table`.
2. The part of `VALUES` is a subquery. It needs to be surround with parentheses.

To fix them, I tried to change the result structure of `DescribeInputRewrite` and `DescribeOutputRewrite` and added test cases for them.

After fixing, we can get the correct SQL:
```sql
SELECT
 "Column Name"
, "Catalog"
, "Schema"
, "Table"
, "Type"
, "Type Size"
, "Aliased"
FROM (
 VALUES 
 ROW ('comment', 'tpch', 'tiny', 'orders', 'varchar(79)', 0, false))
"Statement Output" ("Column Name", "Catalog", "Schema", "Table", "Type", "Type Size", "Aliased")
```

goldmedal commented:
> Thank you for your pull request and welcome to our community. We require contributors to sign our [Contributor License Agreement](https://github.com/trinodb/cla/raw/master/Trino%20Foundation%20Individual%20CLA.pdf), and we don't seem to have you on file. In order for us to review and merge your code, please submit the signed CLA to [cla@trino.io](mailto:cla@trino.io). For more information, see https://github.com/trinodb/cla.

I has sent my signed agreement to your mail. Please check it and many thanks.

kasiafi commented on core/trino-parser/src/main/java/io/trino/sql/QueryUtil.java:
```diff
@@ -162,7 +162,7 @@ public static Relation aliased(Relation relation, String alias, List<String> col
                 identifier(alias),
                 columnAliases.stream()
-                        .map(QueryUtil::identifier)
+                        .map(QueryUtil::quotedIdentifier)
```
There are other sites than `describe`-rewrites involving a `simpleQuery` over `aliased`. They also need to be updated so that the selected column names are quoted.
These are:
- methods `visitShowSession()` and `visitShowFunctions()` in `ShowQueriesRewrite`
- method `rewriteShowStats` in `ShowStatsRewrite`

goldmedal replied:
ok. I'll fix them and add test case for them.

kasiafi commented on core/trino-main/src/test/java/io/trino/sql/rewrite/TestStatementRewrite.java:
```diff
@@ -0,0 +1,210 @@
+        transaction(transactionManager, accessControl)
+                .readUncommitted()
+                .execute(beginTransaction(CLIENT_SESSION_BUILDER), session -> {
+                    Statement root = SQL_PARSER.createStatement("DESCRIBE OUTPUT q1", new ParsingOptions());
```
To make the test more to the point, you could create a `Statement` here instead of invoking the parser.

goldmedal replied:
I don't understand what you said about creating a `Statement`. I used the parser to create a `Statement` from a SQL string actually.
Could you explain more? Thanks.

kasiafi replied:
I meant that you can use a public constructor of `DescribeInput` / `DescribeOutput` to get an instance of those statements instead of parsing them from SQL. Parsing the `DESCRIBE ...` statements is not subject of this change, so no need to make it part of the test.

goldmedal replied:
ok, I got it. I'll fix them. Thank you.

kasiafi commented on core/trino-main/src/test/java/io/trino/sql/rewrite/TestStatementRewrite.java:
```diff
@@ -0,0 +1,210 @@
+                    Statement rewrittenNode = rewrite(session, root, DESCRIBE_OUTPUT_REWRITE);
+                    // SqlFormatterUtil#getFormattedSql will parse query and assert query round-trip inside.
+                    // We don't need to add extra asserting.
+                    SqlFormatterUtil.getFormattedSql(rewrittenNode, SQL_PARSER);
```
The test should include the expected `String` and an assertion that the actual result matches.

goldmedal replied:
ok. I'll add the assertions.

kasiafi commented on core/trino-main/src/test/java/io/trino/sql/rewrite/TestStatementRewrite.java:
```diff
@@ -0,0 +1,210 @@
+                .readUncommitted()
+                .execute(beginTransaction(CLIENT_SESSION_BUILDER), session -> {
+                    Statement root = SQL_PARSER.createStatement("DESCRIBE INPUT q1", new ParsingOptions());
+                    Statement rewrittenNode = rewrite(session, root, DESCRIBE_OUTPUT_REWRITE);
```
I guess this should be `DescribeInputRewrite`.

goldmedal replied:
oops, Yes, it should be `DescribeInputRewrite`. Thanks for the reminder.

kasiafi reviewed (commented):
Thanks for your contribution!

Generally speaking, the reason of the failure you saw is that the AST structure does not quite reflect the SQL grammar. Apparently, the describe statements weren't formatted back to SQL before. Could you please share what led you to do this? I've seen more examples of AST nodes which will fail to roundtrip.

goldmedal commented:
Thanks for your reviewing!

> Could you please share what led you to do this?

Sure. Actually, it's not a common used case. I wanted to print the schema of prepared statements to the file using CTAS statement, and then I found `describe output` can't be use like
```sql
create table c1.q2.output with ( format = 'csv' ) as describe output p1;
```
Then, I tried to understand how `DescribeOutput` work and found it just generates a simple `select` query to create the output. Because I wanted to use CTAS to save the result, I formatted the result of `DescribeOutputRewrite` using `SqlFormatter` and tried to execute it. That's why I found this issue.

martint commented:
@cla-bot check

goldmedal commented:
@kasiafi I've addressed your comments in later commits. Could you take a look? Thanks :)

colebow commented:
:wave: @goldmedal - this PR has become inactive. If you're still interested in working on it, please let us know, and we can try to get reviewers to help with that.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.

goldmedal commented:
> 👋 @goldmedal - this PR has become inactive. If you're still interested in working on it, please let us know, and we can try to get reviewers to help with that.
> 
> We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.

Hi @colebow, I'm not sure whether this PR is worth to do. I had addressed the comments before it having too many merge conflicts. However, the community didn't give me some responses. Maybe it's not a problem for Trino.
I'm still interesting in working on this PR if it's worth. So, let me know if you think the PR should be fixed. I'll try to fix those conflicts.

colebow commented:
@martint - when you get a chance, can you take a look?

@goldmedal I've asked on the Trino Slack and the word I've heard is that this is a real issue which needs to be fixed, it's just sometimes tough to get the wheels spinning on old PRs again.

martint commented:
This is actually not an issue worth fixing. The syntax trees created internally by DescribeInputRewrite and DescribeOutputRewrite are an implementation detail and not meant to be used externally. In the future, we'll likely remove the intermediate syntax tree altogether and implement the describe input/output operations as native IR operations.

As long as DESCRIBE INPUT and DESCRIBE OUTPUT behave as expected, there is not much benefit in adding all the infrastructure for testing the intermediate queries.

goldmedal commented:
I see. If it's only used internally and could be removed in the future, we don't need to fix them now. I agree with @martint on this.
@colebow Please feel free to close this PR. Thanks :)

colebow commented:
Sorry for the back and forth, but glad we've got a concrete resolution!
